pallets/nfts/src/features/approvals.rs

```diff
@@ -165,10 +178,254 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
+						CollectionApprovalCount::<T, I>::mutate(
+							collection,
+							Option::<T::AccountId>::None,
+							|approvals| approvals.saturating_inc(),
+						);
+						Zero::zero()
+					},
+				};
+
+				if current_deposit < deposit_required {
+					T::Currency::reserve(&origin, deposit_required - current_deposit)?;
+					current_deposit = deposit_required;
+				}
+				*maybe_approval = Some((deadline, current_deposit));
+				Ok(())
+			},
+		)?;
+
+		Self::deposit_event(Event::TransferApproved {
+			collection,
+			item: None,
+			owner: origin,
+			delegate,
+			deadline,
+		});
+
+		Ok(())
+	}
+
+	/// Cancels the transfer of items in the collection that owned by the origin to a delegate.
+	///
+	/// This function is used to cancel the approval for the transfer of items in the `collection`
+	/// that owned by the `origin` to a `delegate`. After canceling the approval, the function
+	/// returns the `origin` back the deposited fund and emits the `ApprovalCancelled` event.
+	///
+	/// - `origin`: The account grants permission to cancel the transfer.
+	/// - `collection`: The identifier of the collection.
+	/// - `delegate`: The account that was previously allowed to take control of items in the
+	///   collection that owned by the origin.
+	pub(crate) fn do_cancel_collection_approval(
+		origin: T::AccountId,
+		collection: T::CollectionId,
+		delegate: T::AccountId,
+	) -> DispatchResult {
+		let (_, deposit) = CollectionApprovals::<T, I>::take((&collection, &origin, &delegate))
+			.ok_or(Error::<T, I>::UnknownCollection)?;
+
+		// Update the approval count for the `origin` account and the whole collection.
+		for key in [Some(&origin), Option::<&T::AccountId>::None] {
+			let count = CollectionApprovalCount::<T, I>::get(collection, key)
+				.checked_sub(1)
+				.ok_or(ArithmeticError::Overflow)?;
+			if count == 0 {
+				CollectionApprovalCount::<T, I>::remove(collection, key);
+			} else {
+				CollectionApprovalCount::<T, I>::insert(collection, key, count);
+			}
+		}
+
+		T::Currency::unreserve(&origin, deposit);
+
+		Self::deposit_event(Event::ApprovalCancelled {
+			collection,
+			owner: origin,
+			item: None,
+			delegate,
+		});
+
+		Ok(())
+	}
+
+	/// Clears all collection approvals.
+	///
+	/// This function is used to clear all approvals to transfer items in the `collection` that
+	/// owned by the `origin` to a `delegate`. After clearing all approvals, the function returns
+	/// the `origin` back the deposited fund of each collection approval and emits the
+	/// `AllApprovalsCancelled` event.
+	///
+	/// - `origin`: The account grants permission to clear the transfer.
+	/// - `collection`: The collection ID containing the item.
+	/// - `witness_approvals`: Information on the collection approvals cleared. This must be
+	///   correct.
+	pub(crate) fn do_clear_all_collection_approvals(
+		origin: T::AccountId,
+		collection: T::CollectionId,
+		witness_approvals: u32,
+	) -> Result<u32, DispatchError> {
```

Comment:
Is theis return type necessary? Shouldnt it always be the witness value? If so, no need to track it.